crates/lib/src/qpu/experimental/random.rs

```diff
@@ -0,0 +1,272 @@
+    pub fn new<T: Into<f64>>(
+        destination: String,
+        source: String,
+        sub_region_size: T,
+        seed: String,
+    ) -> Self {
+        Self {
+            destination,
+            source,
+            sub_region_size: Complex64 {
+                re: sub_region_size.into(),
+                im: 0.0,
+            },
+            seed,
+        }
+    }
+}
+
+impl TryFrom<ChooseRandomRealSubRegions> for Call {
+    type Error = CallError;
+
+    fn try_from(value: ChooseRandomRealSubRegions) -> std::result::Result<Self, Self::Error> {
+        Self::try_new(
+            ChooseRandomRealSubRegions::NAME.to_string(),
+            vec![
+                UnresolvedCallArgument::Identifier(value.destination.clone()),
+                UnresolvedCallArgument::Identifier(value.source.to_string()),
+                UnresolvedCallArgument::Immediate(value.sub_region_size),
+                UnresolvedCallArgument::Identifier(value.seed.clone()),
+            ],
+        )
+    }
+}
+
+impl ExternedCall for ChooseRandomRealSubRegions {
+    type Error = Error;
+
+    const NAME: &str = "choose_random_real_sub_regions";
+
+    #[allow(clippy::doc_markdown)]
+    /// Build the signature for the `PRAGMA EXTERN choose_random_real_sub_regions`
+    /// instruction. The signature is:
+    ///
+    /// rust ignore
+    ///     "(destination : mut REAL[], source : REAL[], sub_region_size : INTEGER, seed : mut INTEGER)"
+    fn build_signature() -> Result<quil_rs::instruction::ExternSignature> {
+        use quil_rs::instruction::{ExternParameter, ExternParameterType};
+
+        let parameters = vec![
+            ExternParameter::try_new(
+                "destination".to_string(),
+                true,
+                ExternParameterType::VariableLengthVector(quil_rs::instruction::ScalarType::Real),
+            ),
+            ExternParameter::try_new(
+                "source".to_string(),
+                false,
+                ExternParameterType::VariableLengthVector(quil_rs::instruction::ScalarType::Real),
+            ),
+            ExternParameter::try_new(
+                "sub_region_size".to_string(),
+                false,
+                ExternParameterType::Scalar(quil_rs::instruction::ScalarType::Integer),
+            ),
+            ExternParameter::try_new(
+                "seed".to_string(),
+                true,
+                ExternParameterType::Scalar(quil_rs::instruction::ScalarType::Integer),
+            ),
+        ]
+        .into_iter()
+        .map(|r| r.map_err(Error::from))
+        .collect::<Result<Vec<ExternParameter>>>()?;
+        Ok(quil_rs::instruction::ExternSignature::new(None, parameters))
+    }
+}
+
+/// Hardware values are 48 bits long.
+const MAX_SEQUENCER_VALUE: u64 = 0xFFFF_FFFF_FFFF;
+
+/// Hardware multiplication currently uses the lower 16 bits of
+/// the PRNG value.
+const MAX_UNSIGNED_MULTIPLIER: u64 = 0x0000_0000_FFFF;
```

Comment:
More style nits, but can we move consts to the top of the file? E.g., `MAX_SEQUENCER_VALUE` is used by `Error` but it isn't clear where to look for the definition.

Reply:
While I'm nitpicking (sorry), can we add the highest two bytes to these numbers? I initially read `MAX_SEQUENCER_VALUE` as equivalent to `u64::MAX` because it's all Fs.

```suggestion

/// Hardware values are 48 bits long.
const MAX_SEQUENCER_VALUE: u64 = 0x0000_FFFF_FFFF_FFFF;

/// Hardware multiplication currently uses the lower 16 bits of
/// the PRNG value.
const MAX_UNSIGNED_MULTIPLIER: u64 = 0x0000_0000_0000_FFFF;
```